src/jit/lclvars.cpp

```diff
@@ -1479,15 +1481,106 @@ int __cdecl Compiler::lvaFieldOffsetCmp(const void* field1, const void* field2)
+//
+// Arguments:
+//   lclNum - struct number to try.
+//
+// Return value:
+//   true if the struct var was promoted.
+//
+bool Compiler::StructPromotionHelper::TryPromoteStructVar(unsigned lclNum)
+{
+    if (CanPromoteStructVar(lclNum))
+    {
+#if 0
+            // Often-useful debugging code: if you've narrowed down a struct-promotion problem to a single
+            // method, this allows you to select a subset of the vars to promote (by 1-based ordinal number).
+            static int structPromoVarNum = 0;
+            structPromoVarNum++;
+            if (atoi(getenv("structpromovarnumlo")) <= structPromoVarNum && structPromoVarNum <= atoi(getenv("structpromovarnumhi")))
+#endif // 0
+        if (ShouldPromoteStructVar(lclNum))
+        {
+            PromoteStructVar(lclNum);
+            return true;
+        }
+    }
+    return false;
+}
 
-void Compiler::lvaCanPromoteStructType(CORINFO_CLASS_HANDLE typeHnd, lvaStructPromotionInfo* structPromotionInfo)
+#ifdef DEBUG
+//--------------------------------------------------------------------------------------------
+// CheckFakedType - check that the fldType for this fieldHnd was retyped as requested type.
+//
+// Arguments:
+//   fieldHnd      - the field handler;
+//   requestedType - as which type the field was accessed;
+//
+// Notes:
+//   For example it can happen when such struct A { struct B { long c } } is compiled and we access A.B.c,
+//   it could look like "GT_FIELD struct B.c -> ADDR -> GT_FIELD struct A.B -> ADDR -> LCL_VAR A" , but
+//   "GT_FIELD struct A.B -> ADDR -> LCL_VAR A" can be promoted to "LCL_VAR long A.B" and then
+//   there is type mistmatch between "GT_FIELD struct B.c" and  "LCL_VAR long A.B".
+//
+void Compiler::StructPromotionHelper::CheckFakedType(CORINFO_FIELD_HANDLE fieldHnd, var_types requestedType)
 {
-    assert(eeIsValueClass(typeHnd));
+    assert(fakedFieldsMap.Lookup(fieldHnd));
+    assert(fakedFieldsMap[fieldHnd] == requestedType);
+}
+#endif // DEBUG
 
-    if (typeHnd != structPromotionInfo->typeHnd)
+//--------------------------------------------------------------------------------------------
+// CanPromoteStructType - checks if the struct type can be promoted.
+//
+// Arguments:
+//   typeHnd - struct handle to check.
+//
+// Return value:
+//   true if the struct type can be promoted.
+//
+// Notes:
+//   The last analyzed type is memorized to skip the check if we ask about the same time again next.
+//   However, it was not found profitable to memorize all analyzed types in a map.
+//
+//   The check initializes only nessasary fields in lvaStructPromotionInfo,
+//   so if the promotion is rejected early than the most fields will be uninitialized.
+//
+bool Compiler::StructPromotionHelper::CanPromoteStructType(CORINFO_CLASS_HANDLE typeHnd)
+{
+    assert(compiler->eeIsValueClass(typeHnd));
+    if (structPromotionInfo.typeHnd != typeHnd)
     {
+        // Analyze this type from scratch.
```

Comment:
Why not invert this test and return immediately if the cached value is valid. Then you can de-indent the remainder of the method.

Reply:
Done, thanks, it looks better now.